docs/source/notes/rref.rst

```diff
@@ -0,0 +1,304 @@
+one of Y's ancestors will be alive (in this case, A) and it will
+prevent the owner from deleting the ``OwnerRRef``. More specifically, if the
+owner does not know Y, A cannot be deleted due to **G2**, and the owner knows A
+as the owner is A's parent.
+
+Things get a little trickier if the RRef is created on a user:
+
+
+.. code::
+
+  OwnerRRef
+      ^
+      |
+      A -> Y -> Z
+
+
+If Z calls to_here on the ``UserRRef``, the owner at least knows A when Z is
+deleted, because otherwise, ``to_here`` wouldn't finish. If Z does not call
+to_here, it is possible that the owner receives all messages from Z before
+any message from A and Y. In this case, as the real data of the ``OwnerRRef``
+has not been created yet, there is nothing to be deleted either. It is the same
+as Z does not exist at all. Hence, it's still OK.
+
+Implementation
+--------------
+
+**G1** is implemented by sending out a delete message in ``UserRRef``
+destructor. To provide **G2**, the parent ``UserRRef`` is put into a context
+whenever it is forked, indexed by the new ``ForkId``. The parent ``UserRRef`` is
+only removed from the context when it receives an ACK from the child, and the
+child will only send out the ACK when it is confirmed by the owner.
+
+
+Protocol Scenarios
+^^^^^^^^^^^^^^^^^^
+
+Let's now discuss how the above designs translate to the protocol in four
+scenarios.
+
+User Share RRef with Owner as Return Value
+------------------------------------------
+
+
+.. code::
+  import torch
+  import torch.distributed.rpc as rpc
+
+  # on worker A
+  rref = rpc.remote('B', torch.add, args=(torch.ones(2), 1))
+  # say the rref has RRefId 100 and ForkId 1
+  rref.to_here()
+
+
+In this case, the ``UserRRef`` is created on the user worker A, then it is
+passed to the owner worker B together with the remote message, and then B
+creates the ``OwnerRRef``. The method ``rpc.remote`` returns immediately,
+meaning that the ``UserRRef`` can be forked/used before the owner knows about
+it.
+
+On the owner, when receiving the ``rpc.remote`` call, it will create the
+``OwnerRRef``, and returns an ACK to acknowledge ``{100, 1}``. Only
+after receiving this ACK, can A delete it's ``UserRRef``. This involves both
+**G1** and **G2**. **G1** is obvious. For **G2**, the ``OwnerRRef`` is a child
+of the ``UserRRef``, and the ``UserRRef`` is not deleted until it receives the
+ACK from the owner.
+
+.. image:: https://user-images\.githubusercontent\.com/16999635/69164772-98181300-0abe-11ea-93a7-9ad9f757cd94.png
+    :alt: user_to_owner_ret.png
+    :width: 500 px
+
+The diagram above shows the message flow, where solid arrow contains user
+function and dashed arrow are builtin messages. Note that the first two messages
+from A to B (``remote`` and ``to_here``) may arrive at B in any order, but the
+final delete message will only be sent out when:
+
+- B acknowledges ``UserRRef {100, 1}`` (G2), and
+- Python GC agrees to delete the local ``UserRRef`` instance.
```

Comment:
Clarify that this happens when the rref is no longer in scope and it is eligible for garbage collection?